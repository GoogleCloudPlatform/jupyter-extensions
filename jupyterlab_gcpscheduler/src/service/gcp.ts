/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable @typescript-eslint/camelcase */
import { ServerConnection } from '@jupyterlab/services';

import {
  CLOUD_FUNCTION_NAME,
  CLOUD_FUNCTION_REGION,
  IMPORT_DIRECTORY,
} from '../data';
import { ProjectStateService } from './project_state';
import {
  handleApiError,
  TransportService,
  GET,
  POST,
} from 'gcp_jupyterlab_shared';
import { IDocumentManager } from '@jupyterlab/docmanager';
import { Notebook } from '@jupyterlab/notebook';

const SERVICE_ACCOUNT_DOMAIN = 'appspot.gserviceaccount.com';
const IMMEDIATE_JOB_INDICATOR = 'jupyterlab_immediate_notebook';
const SCHEDULED_JOB_INDICATOR = 'jupyterlab_scheduled_notebook';

/**
 * Cloud Scheduler Job
 * https://cloud.google.com/scheduler/docs/reference/rest/v1/projects.locations.jobs#Job
 */
export interface CloudSchedulerJob {
  name: string;
  description: string;
  schedule: string;
  timeZone: string;
  httpTarget: {
    body: string;
    headers: { [name: string]: string };
    httpMethod: string;
    uri: string;
    oidcToken: { serviceAccountEmail: string };
  };
}

/** Message type describing an AI Platform training Job */
export interface RunNotebookRequest {
  imageUri: string;
  inputNotebookGcsPath: string;
  jobId: string;
  masterType: string;
  outputNotebookGcsPath: string;
  scaleTier: string;
  region: string;
  acceleratorType: string;
  acceleratorCount: string;
}

/** List of Jobs returned from AI Platform. */
export type ListAiPlatformJobsResponse = gapi.client.ml.GoogleCloudMlV1__ListJobsResponse;

/** AI Platform Job. */
export type AiPlatformJob = gapi.client.ml.GoogleCloudMlV1__Job;

type StorageObject = gapi.client.storage.Object;

const AI_PLATFORM = 'https://ml.googleapis.com/v1';
const CLOUD_SCHEDULER = 'https://cloudscheduler.googleapis.com/v1';
const STORAGE_DOWNLOAD = 'https://storage.googleapis.com/storage/v1';
const STORAGE_UPLOAD = 'https://storage.googleapis.com/upload/storage/v1';
/**
 * Class to interact with GCP services.
 */
export class GcpService {
  private readonly serverSettings = ServerConnection.defaultSettings;
  private readonly runtimeUrl = `${this.serverSettings.baseUrl}gcp/v1/runtime`;

  constructor(
    private _transportService: TransportService,
    private _projectStateService: ProjectStateService,
    private _documentManager: IDocumentManager
  ) {}

  get projectId() {
    return this._projectStateService.projectId;
  }

  set transportService(transportService: TransportService) {
    this._transportService = transportService;
  }

  /**
   * Uploads the specified Notebook JSON representation to the specified path.
   */
  async uploadNotebook(
    notebookContents: string,
    gcsPath: string
  ): Promise<StorageObject> {
    try {
      const { bucket, object } = this._getGcsPathParts(gcsPath);
      const response = await this._transportService.submit<StorageObject>({
        path: `${STORAGE_UPLOAD}/b/${bucket}/o`,
        method: POST,
        headers: { 'Content-Type': 'application/json' },
        body: notebookContents,
        params: {
          name: object,
          uploadType: 'media',
        },
      });
      return response.result;
    } catch (err) {
      console.error(`Unable to upload Notebook contents to ${gcsPath}`);
      handleApiError(err);
    }
  }

  /**
   * Downloads the notebook located at the gcsPath provided and retrieves the file name from the
   * JSON. Creates a new import directory and attempts to rename it, if it fails then creates
   * the notebook in the already-existing directory and deletes the new directory.
   *
   * @param gcsPath Path to the notebook we want to create.
   */
  async importNotebook(gcsPath: string): Promise<Notebook> {
    const contents = await this.downloadNotebook(gcsPath);

    const retrievedName = JSON.parse(contents)
      .metadata?.papermill?.input_path?.split('/')
      .slice(-1)[0];
    const fileName = retrievedName ? retrievedName : 'untitled.ipynb';

    const newDirectory = await this._documentManager.newUntitled({
      type: 'directory',
    });

    try {
      await this._documentManager.rename(newDirectory.path, IMPORT_DIRECTORY);
    } catch (err) {
      // Expect a 409 error if directory exists. Otherwise throw the error
      if (err.message.indexOf('409') === -1) {
        throw err;
      }
      this._documentManager.deleteFile(newDirectory.path);
    }

    return this.createNotebook(fileName, contents);
  }

  private createNotebook(fileName: string, contents: string): Notebook {
    const widget = this._documentManager.createNew(
      IMPORT_DIRECTORY + fileName,
      'Notebook'
    ) as Notebook;
    widget.model.fromString(contents);
    return widget;
  }

  async downloadNotebook(gcsPath: string): Promise<string> {
    try {
      const { bucket, object } = this._getGcsPathParts(gcsPath);
      const response = await this._transportService.submit<string>({
        path: `${STORAGE_DOWNLOAD}/b/${bucket}/o/${encodeURIComponent(object)}`,
        params: {
          alt: 'media',
        },
      });
      return typeof response.result === 'string'
        ? response.result
        : JSON.stringify(response.result);
    } catch (err) {
      console.error(`Unable to download Notebook contents at ${gcsPath}`);
      handleApiError(err);
    }
  }

  /**
   * Submits a Notebook for recurring scheduled execution on AI Platform via a
   * new Cloud Scheduler job.
   * @param request
   * @param regionName
   * @param serviceAccountEmail
   * @param schedule
   */
  async scheduleNotebook(
    request: RunNotebookRequest,
    regionName: string,
    schedule: string
  ): Promise<CloudSchedulerJob> {
    let timeZone = 'America/New_York';
    try {
      timeZone = Intl.DateTimeFormat().resolvedOptions().timeZone;
    } catch (err) {
      console.warn('Unable to determine timezone');
    }

    try {
      const projectId = await this.projectId;
      const locationPrefix = `projects/${projectId}/locations/${regionName}/jobs`;
      const requestBody: CloudSchedulerJob = {
        name: `${locationPrefix}/${request.jobId}`,
        description: SCHEDULED_JOB_INDICATOR,
        schedule,
        timeZone,
        httpTarget: {
          body: btoa(
            JSON.stringify(this._buildAiPlatformJobRequest(request, true))
          ),
          headers: { 'Content-Type': 'application/json' },
          httpMethod: POST,
          oidcToken: {
            serviceAccountEmail: `${projectId}@${SERVICE_ACCOUNT_DOMAIN}`,
          },
          uri: `https://${CLOUD_FUNCTION_REGION}-${projectId}.cloudfunctions.net/${CLOUD_FUNCTION_NAME}`,
        },
      };
      const response = await this._transportService.submit<CloudSchedulerJob>({
        path: `${CLOUD_SCHEDULER}/${locationPrefix}`,
        method: POST,
        body: requestBody,
      });
      return response.result;
    } catch (err) {
      console.error('Unable to create Cloud Scheduler job');
      handleApiError(err);
    }
  }

  /**
   * Submits a Notebook for immediate execution on AI Platform.
   * @param cloudFunctionUrl
   * @param request
   */
  async runNotebook(request: RunNotebookRequest): Promise<AiPlatformJob> {
    try {
      const projectId = await this.projectId;
      const response = await this._transportService.submit<AiPlatformJob>({
        path: `${AI_PLATFORM}/projects/${projectId}/jobs`,
        method: POST,
        body: this._buildAiPlatformJobRequest(request),
      });
      return response.result;
    } catch (err) {
      console.error('Unable to submit Notebook to AI Platform');
      handleApiError(err);
    }
  }

  /**
   * Gets list of AiPlatformJobs
   * @param cloudFunctionUrl
   * @param request
   */
  async listNotebookJobs(
    pageToken?: string
  ): Promise<ListAiPlatformJobsResponse> {
    try {
      const filter = [SCHEDULED_JOB_INDICATOR, IMMEDIATE_JOB_INDICATOR]
        .map(v => `labels.job_type=${v}`)
        .join(' OR ');
      const projectId = await this.projectId;
      const params: { [k: string]: string } = { filter };
      if (pageToken) {
        params['pageToken'] = pageToken;
      }
      const response = await this._transportService.submit<
        ListAiPlatformJobsResponse
      >({
        path: `${AI_PLATFORM}/projects/${projectId}/jobs`,
        params,
      });
      return response.result;
    } catch (err) {
      console.error('Unable to list AI Platform Notebook jobs');
      handleApiError(err);
    }
  }

  async getImageUri(): Promise<string> {
    // TODO need to check if image exist
    const imageUriPrefix = 'gcr.io/deeplearning-platform-release/';
    const runtimeEnv = await this._getRuntimeEnv();
    if (!runtimeEnv) return '';

    const lastDotIndex = runtimeEnv.lastIndexOf('.');
    return `${imageUriPrefix}${runtimeEnv.substr(0, lastDotIndex)}`;
  }

  private async _getRuntimeEnv(): Promise<string> {
    try {
      const response = await ServerConnection.makeRequest(
        this.runtimeUrl,
        { method: GET },
        this.serverSettings
      );
      return await response.text();
    } catch (err) {
      console.warn('Unable to obtain runtime environment');
      return '';
    }
  }

  private _buildAiPlatformJobRequest(
    request: RunNotebookRequest,
    isScheduled = false
  ): AiPlatformJob {
    return {
      jobId: request.jobId,
      labels: {
        job_type: isScheduled
          ? SCHEDULED_JOB_INDICATOR
          : IMMEDIATE_JOB_INDICATOR,
      },
      trainingInput: {
        args: [
          'nbexecutor',
          '--input-notebook',
          request.inputNotebookGcsPath,
          '--output-notebook',
          request.outputNotebookGcsPath,
        ],
        masterConfig: {
          imageUri: request.imageUri,
          acceleratorConfig: {
            count: request.acceleratorCount || undefined,
            type: request.acceleratorType || undefined,
          },
        },
        masterType: request.masterType || undefined,
        region: request.region,
        scaleTier: request.scaleTier,
      },
    } as AiPlatformJob;
  }

  private _getGcsPathParts(
    gcsPath: string
  ): { bucket: string; object: string } {
    if (gcsPath.startsWith('gs://')) {
      gcsPath = gcsPath.slice(5);
    }
    const pathParts = gcsPath.split('/');
    return {
      bucket: pathParts[0],
      object: pathParts.slice(1).join('/'),
    };
  }
}
