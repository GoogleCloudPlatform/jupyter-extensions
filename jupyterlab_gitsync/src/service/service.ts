import { FileTracker } from './tracker';
import { GitManager } from './git';
import { ISignal, Signal } from '@lumino/signaling';

/**
 *
 * Class in charge of calling necessary functions from
 * the GitManager and FileTracker classes
 *
 */

export class GitSyncService {
  /* Member Fields */
  private _git: GitManager;
  private _tracker: FileTracker;
  private _completed: boolean = true;
  private _running: boolean;
  syncInterval: number = 0 * 1000;

  private _status: 'sync' | 'merge' | 'up-to-date' | 'dirty' | 'warning';
  private _blocked: boolean = false;
  private _stateChange: Signal<this, boolean> = new Signal<this, boolean>(this);
  private _statusChange: Signal<this, string> = new Signal<this, string>(this);

  constructor(git: GitManager, tracker: FileTracker) {
    this._git = git;
    this._tracker = tracker;
    this._addListeners();
  }

  start() {
    if (!this._blocked && !this.running){
      console.log('start git sync service');
      this._running = true;
      this._run();
      this._stateChange.emit(this.running);
    }
  }

  stop() {
    if (!this._blocked && this.running){
      this._running = false;
      this._stateChange.emit(this.running);
      console.log('stop git sync service');
    }
  }

  get running(): boolean {
    return this._running;
  }

  get completed(): boolean {
    return this._completed;
  }

  get git(): GitManager {
    return this._git;
  }

  get tracker(): FileTracker {
    return this._tracker;
  }

  get status(): string {
    return this._status;
  }

  get stateChange(): ISignal<this, boolean> {
    return this._stateChange;
  }

  get statusChange(): ISignal<this, string> {
    return this._statusChange;
  }

  private _updateStatus(status: 'sync' | 'merge' | 'up-to-date' | 'dirty' | 'warning'){
    if (status != this.status){
      this._status = status;
      this._statusChange.emit(status);
    }
  }

  private async _run(): Promise<void> {
    if (this.running && this.completed){
      this._completed = false;
      setTimeout(async () => {
        try{
          await this.tracker.saveAll();
          this._updateStatus('sync')
          await this.git.sync();
          this._updateStatus('merge')
          await this.tracker.reloadAll();
          this._updateStatus('up-to-date');
        } catch (error) {
          console.warn(error);
          this._updateStatus('warning')
        }
        this._completed = true; 
        this._run();
      }, this.syncInterval);
    }
  }

  private _addListener(signal: ISignal<any, any>, callback: any){
    return signal.connect(callback, this);
  }

  private _conflictListener(sender: FileTracker, conflict: boolean){
    if (!conflict && !this.running){
      this._blocked = false;
      this.start();
    } else if (conflict && this.running){
      this.stop();
      this._blocked = true;
    }
  }

  private _dirtyStateListener(sender: FileTracker, dirty: boolean){
    if (this.status === 'up-to-date' && dirty){
      this._updateStatus('dirty');
    }
  }

  private _addListeners(): void {
    this._addListener(this.tracker.conflictState, this._conflictListener);
    this._addListener(this.tracker.dirtyState, this._dirtyStateListener);
  }
}
