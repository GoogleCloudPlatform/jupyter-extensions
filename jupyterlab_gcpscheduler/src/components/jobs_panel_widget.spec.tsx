/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable @typescript-eslint/no-unused-vars */
import { shallow } from 'enzyme';
import * as React from 'react';

import {
  GcpScheduledJobsWidget,
  GcpScheduledJobsPanel,
} from './jobs_panel_widget';
import { GcpService } from '../service/gcp';
import { TEST_PROJECT, getExecution, getSchedule } from '../test_helpers';

describe('GcpScheduledJobsWidget', () => {
  const fakeGcpService = {} as GcpService;

  it('Renders with undefined visibility status', () => {
    const widget = new GcpScheduledJobsWidget(fakeGcpService);
    const element = shallow(widget.render());
    expect(
      element.find(GcpScheduledJobsPanel).prop('isVisible')
    ).toBeUndefined();
  });

  it('Set isVisible based on showing and hiding', () => {
    const widget = new GcpScheduledJobsWidget(fakeGcpService);
    const element = shallow(widget.render());
    widget.onAfterShow();
    expect(element.find(GcpScheduledJobsPanel).prop('isVisible')).toBe(true);

    widget.onAfterHide();
    expect(element.find(GcpScheduledJobsPanel).prop('isVisible')).toBe(false);
  });
});

describe('GcpScheduledJobsPanel', () => {
  const mockProjectId = jest.fn();
  const mockListExecutions = jest.fn();
  const mockListSchedules = jest.fn();
  const mockGcpService = ({
    get projectId() {
      return mockProjectId();
    },
    listExecutions: mockListExecutions,
    listSchedules: mockListSchedules,
  } as unknown) as GcpService;

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('Renders when not visible', () => {
    const component = shallow(
      <GcpScheduledJobsPanel gcpService={mockGcpService} isVisible={false} />
    );
    expect(component).toMatchSnapshot();
    expect(mockListExecutions).not.toHaveBeenCalled();
    expect(mockListSchedules).not.toHaveBeenCalled();
  });

  it('Shows error if project cannot be determined', async () => {
    const rejectedProjectId = Promise.reject('No project');
    mockProjectId.mockReturnValue(rejectedProjectId);
    const component = shallow(
      <GcpScheduledJobsPanel gcpService={mockGcpService} isVisible={false} />
    );
    await rejectedProjectId.catch(() => null);

    expect(component).toMatchSnapshot();
    expect(mockListExecutions).not.toHaveBeenCalled();
    expect(mockListSchedules).not.toHaveBeenCalled();
  });

  it('Shows error if executions cannot be retrieved', async () => {
    const rejectedPromise = Promise.reject('Cannot retrieve executions');
    const resolvedSchedules = Promise.resolve({
      schedules: [
        getSchedule(),
        { ...getSchedule(), id: 'notebook_execution2_fghijk' },
      ],
    });
    mockProjectId.mockResolvedValue(TEST_PROJECT);
    mockListExecutions.mockReturnValue(rejectedPromise);
    mockListSchedules.mockResolvedValue(resolvedSchedules);

    const component = shallow(
      <GcpScheduledJobsPanel gcpService={mockGcpService} isVisible={false} />
    );
    component.setProps({ isVisible: true });
    await rejectedPromise.catch(() => null);
    await expect(component).toMatchSnapshot();
    expect(mockListExecutions).toHaveBeenCalled();
    expect(mockListSchedules).not.toHaveBeenCalled();
  });

  it('Shows error if schedules cannot be retrieved', async () => {
    const rejectedPromise = Promise.reject('Cannot retrieve schedules');
    const resolvedExecutions = Promise.resolve({
      executions: [
        getExecution(),
        { ...getExecution(), id: 'notebook_execution2_fghijk' },
      ],
    });
    mockProjectId.mockResolvedValue(TEST_PROJECT);
    mockListExecutions.mockResolvedValue(resolvedExecutions);
    mockListSchedules.mockReturnValue(rejectedPromise);

    const component = shallow(
      <GcpScheduledJobsPanel gcpService={mockGcpService} isVisible={false} />
    );
    component.setProps({ isVisible: true });
    await resolvedExecutions;

    (component.instance() as GcpScheduledJobsPanel).handleChangeTab(null, 1);
    await rejectedPromise.catch(() => null);

    await expect(component).toMatchSnapshot();
    expect(component.state('tab')).toEqual(1);
    expect(mockListExecutions).toHaveBeenCalled();
    expect(mockListSchedules).toHaveBeenCalled();
  });

  it('Shows loading indicator', async () => {
    const resolvedProject = Promise.resolve(TEST_PROJECT);
    const resolvedExecutions = Promise.resolve({
      executions: [
        getExecution(),
        { ...getExecution(), id: 'notebook_execution2_fghijk' },
      ],
    });
    const resolvedSchedules = Promise.resolve({
      schedules: [
        getSchedule(),
        { ...getSchedule(), id: 'notebook_execution2_fghijk' },
      ],
    });
    mockProjectId.mockReturnValue(resolvedProject);
    mockListExecutions.mockReturnValue(resolvedExecutions);
    mockListSchedules.mockReturnValue(resolvedSchedules);

    const component = shallow(
      <GcpScheduledJobsPanel gcpService={mockGcpService} isVisible={false} />
    );
    component.setProps({ isVisible: true });

    expect(component).toMatchSnapshot();
    expect(mockListExecutions).toHaveBeenCalled();
    expect(mockListSchedules).not.toHaveBeenCalled();
  });

  it('Shows executions', async () => {
    const resolvedExecutions = Promise.resolve({
      executions: [
        getExecution(),
        { ...getExecution(), id: 'notebook_execution2_fghijk' },
      ],
    });
    const resolvedSchedules = Promise.resolve({
      schedules: [
        getSchedule(),
        { ...getSchedule(), id: 'notebook_execution2_fghijk' },
      ],
    });
    mockProjectId.mockResolvedValue(TEST_PROJECT);
    mockListExecutions.mockReturnValue(resolvedExecutions);
    mockListSchedules.mockReturnValue(resolvedSchedules);

    const component = shallow(
      <GcpScheduledJobsPanel gcpService={mockGcpService} isVisible={false} />
    );
    component.setProps({ isVisible: true });
    await resolvedExecutions;
    // await resolvedSchedules;

    expect(component).toMatchSnapshot();
    expect(mockListExecutions).toHaveBeenCalled();
    expect(mockListSchedules).not.toHaveBeenCalled();
  });

  it('Shows schedules', async () => {
    const resolvedExecutions = Promise.resolve({
      executions: [
        getExecution(),
        { ...getExecution(), id: 'notebook_execution2_fghijk' },
      ],
    });
    const resolvedSchedules = Promise.resolve({
      schedules: [
        getSchedule(),
        { ...getSchedule(), id: 'notebook_execution2_fghijk' },
      ],
    });
    mockProjectId.mockResolvedValue(TEST_PROJECT);
    mockListExecutions.mockReturnValue(resolvedExecutions);
    mockListSchedules.mockReturnValue(resolvedSchedules);

    const component = shallow(
      <GcpScheduledJobsPanel gcpService={mockGcpService} isVisible={false} />
    );
    component.setProps({ isVisible: true });
    await resolvedExecutions;

    (component.instance() as GcpScheduledJobsPanel).handleChangeTab(null, 1);
    await resolvedSchedules;

    expect(component).toMatchSnapshot();
    expect(component.state('tab')).toEqual(1);
    expect(mockListExecutions).toHaveBeenCalled();
    expect(mockListSchedules).toHaveBeenCalled();
  });

  it('Refreshes executions', async () => {
    const resolvedExecutions = Promise.resolve({
      executions: [
        getExecution(),
        { ...getExecution(), id: 'notebook_execution2_fghijk' },
      ],
    });
    mockProjectId.mockResolvedValue(TEST_PROJECT);
    mockListExecutions.mockReturnValue(resolvedExecutions);

    const component = shallow(
      <GcpScheduledJobsPanel gcpService={mockGcpService} isVisible={false} />
    );
    component.setProps({ isVisible: true });
    await resolvedExecutions;

    component.find({ title: 'Refresh' }).simulate('click');
    expect(mockListExecutions).toHaveBeenCalledTimes(2);
  });

  it('Executions pagination works', async () => {
    const resolvedExecutions = Promise.resolve({
      executions: [
        getExecution(),
        { ...getExecution(), id: 'notebook_execution2_fghijk' },
      ],
      pageToken: 'abc',
    });
    const nextPageResolvedExecutions = Promise.resolve({
      executions: [
        getExecution(),
        { ...getExecution(), id: 'notebook_execution2_fghijk' },
      ],
      pageToken: 'def',
    });
    const lastPageResolvedExecutions = Promise.resolve({
      executions: [
        getExecution(),
        { ...getExecution(), id: 'notebook_execution2_fghijk' },
      ],
      pageToken: 'ghi',
    });
    mockProjectId.mockResolvedValue(TEST_PROJECT);
    mockListExecutions
      .mockReturnValueOnce(resolvedExecutions)
      .mockReturnValueOnce(nextPageResolvedExecutions)
      .mockReturnValueOnce(lastPageResolvedExecutions);
    const component = shallow(
      <GcpScheduledJobsPanel gcpService={mockGcpService} isVisible={false} />
    );
    component.setProps({ isVisible: true });
    await resolvedExecutions;
    expect(component.state('page')).toEqual(0);
    (component.instance() as GcpScheduledJobsPanel).handleChangePage(null, 1);
    expect(component.state('page')).toEqual(1);
    await nextPageResolvedExecutions;
    expect(mockListExecutions).toHaveBeenCalledWith('', 10, 'abc');
    (component.instance() as GcpScheduledJobsPanel).handleChangePage(null, 2);
    expect(component.state('page')).toEqual(2);
    await lastPageResolvedExecutions;
    expect(mockListExecutions).toHaveBeenCalledWith('', 10, 'def');
    expect(mockListExecutions).toHaveBeenCalledTimes(3);
  });

  it('Executions page size change', async () => {
    const resolvedExecutions = Promise.resolve({
      executions: [
        getExecution(),
        { ...getExecution(), id: 'notebook_execution2_fghijk' },
      ],
      pageToken: 'abc',
    });
    mockProjectId.mockResolvedValue(TEST_PROJECT);
    mockListExecutions.mockReturnValue(resolvedExecutions);
    const component = shallow(
      <GcpScheduledJobsPanel gcpService={mockGcpService} isVisible={false} />
    );
    component.setProps({ isVisible: true });
    await resolvedExecutions;
    expect(component.state('rowsPerPage')).toEqual(10);
    (component.instance() as GcpScheduledJobsPanel).handleChangeRowsPerPage({
      target: {
        value: '100',
      },
    } as React.ChangeEvent<HTMLInputElement>);
    expect(component.state('rowsPerPage')).toEqual(100);
    expect(mockListExecutions).toHaveBeenCalledTimes(2);
  });

  it('Schedules pagination works', async () => {
    const resolvedExecutions = Promise.resolve({
      executions: [
        getExecution(),
        { ...getExecution(), id: 'notebook_execution2_fghijk' },
      ],
    });
    const resolvedSchedules = Promise.resolve({
      schedules: [
        getSchedule(),
        { ...getSchedule(), id: 'notebook_execution2_fghijk' },
      ],
      pageToken: 'abc',
    });
    const nextPageResolvedSchedules = Promise.resolve({
      schedules: [
        getSchedule(),
        { ...getSchedule(), id: 'notebook_execution2_fghijk' },
      ],
      pageToken: 'def',
    });
    const lastPageResolvedSchedules = Promise.resolve({
      schedules: [
        getSchedule(),
        { ...getSchedule(), id: 'notebook_execution2_fghijk' },
      ],
      pageToken: 'ghi',
    });
    mockProjectId.mockResolvedValue(TEST_PROJECT);
    mockListExecutions.mockReturnValue(resolvedExecutions);
    mockListSchedules
      .mockReturnValueOnce(resolvedSchedules)
      .mockReturnValueOnce(nextPageResolvedSchedules)
      .mockReturnValueOnce(lastPageResolvedSchedules);
    const component = shallow(
      <GcpScheduledJobsPanel gcpService={mockGcpService} isVisible={false} />
    );
    component.setProps({ isVisible: true });
    await resolvedExecutions;

    (component.instance() as GcpScheduledJobsPanel).handleChangeTab(null, 1);
    await resolvedSchedules;

    expect(component.state('tab')).toEqual(1);
    expect(component.state('page')).toEqual(0);
    (component.instance() as GcpScheduledJobsPanel).handleChangePage(null, 1);
    expect(component.state('page')).toEqual(1);
    await nextPageResolvedSchedules;
    expect(mockListSchedules).toHaveBeenCalledWith(10, 'abc');
    (component.instance() as GcpScheduledJobsPanel).handleChangePage(null, 2);
    expect(component.state('page')).toEqual(2);
    await lastPageResolvedSchedules;
    expect(mockListSchedules).toHaveBeenCalledWith(10, 'def');
    expect(mockListSchedules).toHaveBeenCalledTimes(3);
  });

  it('Schedules page size change', async () => {
    const resolvedExecutions = Promise.resolve({
      executions: [
        getExecution(),
        { ...getExecution(), id: 'notebook_execution2_fghijk' },
      ],
    });
    const resolvedSchedules = Promise.resolve({
      schedules: [
        getSchedule(),
        { ...getSchedule(), id: 'notebook_execution2_fghijk' },
      ],
      pageToken: 'abc',
    });
    mockProjectId.mockResolvedValue(TEST_PROJECT);
    mockListExecutions.mockReturnValue(resolvedExecutions);
    mockListSchedules.mockReturnValue(resolvedSchedules);
    const component = shallow(
      <GcpScheduledJobsPanel gcpService={mockGcpService} isVisible={false} />
    );
    component.setProps({ isVisible: true });
    await resolvedExecutions;

    (component.instance() as GcpScheduledJobsPanel).handleChangeTab(null, 1);
    await resolvedSchedules;

    expect(component.state('tab')).toEqual(1);
    expect(component.state('rowsPerPage')).toEqual(10);
    (component.instance() as GcpScheduledJobsPanel).handleChangeRowsPerPage({
      target: {
        value: '100',
      },
    } as React.ChangeEvent<HTMLInputElement>);
    expect(component.state('rowsPerPage')).toEqual(100);
    expect(mockListSchedules).toHaveBeenCalledTimes(2);
  });
});
