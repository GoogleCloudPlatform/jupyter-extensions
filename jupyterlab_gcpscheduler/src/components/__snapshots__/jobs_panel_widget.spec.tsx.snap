// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GcpScheduledJobsPanel Renders when not visible 1`] = `
<div
  className="panel_f16v76lc"
>
  <div
    className="headerContainer_f17cdueg"
  >
    <header
      className="header_f10v7dh6"
    >
      AI Platform Notebook Jobs
    </header>
    <WithStyles(WithStyles(ForwardRef(IconButton)))
      onClick={[Function]}
      title="Refresh Jobs"
    >
      <WithStyles(RefreshIcon) />
    </WithStyles(WithStyles(ForwardRef(IconButton)))>
  </div>
  <ul
    className="list_fqi18ru"
  />
</div>
`;

exports[`GcpScheduledJobsPanel Shows error if jobs cannot be retrieved 1`] = `
<div
  className="panel_f16v76lc"
>
  <div
    className="headerContainer_f17cdueg"
  >
    <header
      className="header_f10v7dh6"
    >
      AI Platform Notebook Jobs
    </header>
    <WithStyles(WithStyles(ForwardRef(IconButton)))
      onClick={[Function]}
      title="Refresh Jobs"
    >
      <WithStyles(RefreshIcon) />
    </WithStyles(WithStyles(ForwardRef(IconButton)))>
  </div>
  <Message
    asError={true}
    text="Cannot retrieve jobs: Unable to retrieve Jobs"
  />
</div>
`;

exports[`GcpScheduledJobsPanel Shows error if project cannot be determined 1`] = `
<div
  className="panel_f16v76lc"
>
  <div
    className="headerContainer_f17cdueg"
  >
    <header
      className="header_f10v7dh6"
    >
      AI Platform Notebook Jobs
    </header>
    <WithStyles(WithStyles(ForwardRef(IconButton)))
      onClick={[Function]}
      title="Refresh Jobs"
    >
      <WithStyles(RefreshIcon) />
    </WithStyles(WithStyles(ForwardRef(IconButton)))>
  </div>
  <Message
    asError={true}
    text="No project: Unable to determine GCP project"
  />
</div>
`;

exports[`GcpScheduledJobsPanel Shows jobs 1`] = `
<div
  className="panel_f16v76lc"
>
  <div
    className="headerContainer_f17cdueg"
  >
    <header
      className="header_f10v7dh6"
    >
      AI Platform Notebook Jobs
    </header>
    <WithStyles(WithStyles(ForwardRef(IconButton)))
      onClick={[Function]}
      title="Refresh Jobs"
    >
      <WithStyles(RefreshIcon) />
    </WithStyles(WithStyles(ForwardRef(IconButton)))>
  </div>
  <ul
    className="list_fqi18ru"
  >
    <JobListItem
      gcpService={
        Object {
          "listNotebookJobs": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
          "projectId": Promise {},
        }
      }
      job={
        Object {
          "createTime": "2020-05-01T19:00:07Z",
          "endTime": "2020-05-01T19:09:42Z",
          "jobId": "notebook_job1_abcxyz",
          "labels": Object {
            "job_type": "jupyterlab_scheduled_notebook",
            "scheduler_job_name": "notebook_job1",
          },
          "startTime": "2020-05-01T19:04:08Z",
          "state": "SUCCEEDED",
          "trainingInput": Object {
            "args": Array [
              "nbexecutor",
              "--input-notebook",
              "gs://test-project/notebook_job1/nb.ipynb",
              "--output-notebook",
              "gs://test-project/notebook_job1/job1.ipynb",
            ],
            "masterConfig": Object {
              "imageUri": "gcr.io/deeplearning-platform-release/tf-cpu.1-15:latest",
            },
            "region": "us-central1",
          },
          "trainingOutput": Object {
            "consumedMLUnits": 0.06,
          },
        }
      }
      key="notebook_job1_abcxyz"
    />
    <JobListItem
      gcpService={
        Object {
          "listNotebookJobs": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
          "projectId": Promise {},
        }
      }
      job={
        Object {
          "createTime": "2020-05-01T19:00:07Z",
          "endTime": "2020-05-01T19:09:42Z",
          "jobId": "notebook_job2_fghijk",
          "labels": Object {
            "job_type": "jupyterlab_scheduled_notebook",
            "scheduler_job_name": "notebook_job1",
          },
          "startTime": "2020-05-01T19:04:08Z",
          "state": "SUCCEEDED",
          "trainingInput": Object {
            "args": Array [
              "nbexecutor",
              "--input-notebook",
              "gs://test-project/notebook_job1/nb.ipynb",
              "--output-notebook",
              "gs://test-project/notebook_job1/job1.ipynb",
            ],
            "masterConfig": Object {
              "imageUri": "gcr.io/deeplearning-platform-release/tf-cpu.1-15:latest",
            },
            "region": "us-central1",
          },
          "trainingOutput": Object {
            "consumedMLUnits": 0.06,
          },
        }
      }
      key="notebook_job2_fghijk"
    />
  </ul>
</div>
`;

exports[`GcpScheduledJobsPanel Shows loading indicator 1`] = `
<div
  className="panel_f16v76lc"
>
  <div
    className="headerContainer_f17cdueg"
  >
    <header
      className="header_f10v7dh6"
    >
      AI Platform Notebook Jobs
    </header>
    <WithStyles(WithStyles(ForwardRef(IconButton)))
      onClick={[Function]}
      title="Refresh Jobs"
    >
      <WithStyles(RefreshIcon) />
    </WithStyles(WithStyles(ForwardRef(IconButton)))>
  </div>
  <WithStyles(ForwardRef(LinearProgress)) />
</div>
`;
